test/cusparse/interfaces.jl

```diff
@@ -93,126 +93,211 @@ using LinearAlgebra, SparseArrays
+                @test c ≈ collect(dc)
+            end
+        end
+
+        @testset "$SparseMatrixType * CuVector -- mul!(c, A, b) $elty" for elty in [Float32, Float64, ComplexF32, ComplexF64]
+            for opa in (identity, transpose, adjoint)
+                CUSPARSE.version() < v"12.0" && SparseMatrixType == CuSparseMatrixCSC && elty <: Complex && opa == adjoint && continue
+                n = 10
+                m = 20
+                alpha = rand(elty)
+                beta = rand(elty)
+                A = opa == identity ? sprand(elty, n, m, 0.2) : sprand(elty, m, n, 0.2)
+                b = rand(elty, m)
+                c = rand(elty, n)
+
+                dA = SparseMatrixType(A)
+                db = CuArray(b)
+                dc = CuArray(c)
+
+                mul!(c, opa(A), b, alpha, beta)
+                mul!(dc, opa(dA), db, alpha, beta)
+                @test c ≈ collect(dc)
+            end
+        end
 
-        A = A + transpose(A)
-        dA = CuSparseMatrixCSR(A)
+        @testset "$SparseMatrixType * CuMatrix -- A * B $elty" for elty in [Float32, Float64, ComplexF32, ComplexF64]
+            for opa in (identity, transpose, adjoint)
+                for opb in (identity, transpose, adjoint)
+                    CUSPARSE.version() < v"12.0" && SparseMatrixType == CuSparseMatrixCSC && elty <: Complex && opa == adjoint && continue
+                    n = 10
+                    k = 15
+                    m = 20
+                    A = opa == identity ? sprand(elty, m, k, 0.2) : sprand(elty, k, m, 0.2)
+                    B = opb == identity ? rand(elty, k, n) : rand(elty, n, k)
 
-        @assert issymmetric(A)
-        mul!(C, f(Symmetric(A)), h(B), alpha, beta)
-        mul!(dC, f(Symmetric(dA)), h(dB), alpha, beta)
-        @test C ≈ collect(dC)
-    end
+                    dA = SparseMatrixType(A)
+                    dB = CuArray(B)
 
-    @testset "issue #1095 ($elty)" for elty in [Float32, Float64, ComplexF32, ComplexF64]
-        # Test non-square matrices
-        n, m, p = 10, 20, 4
-        alpha = rand()
-        beta = rand()
-        A = sprand(elty, n, m, rand())
-        B = rand(elty, m, p)
-        C = rand(elty, n, p)
-
-        dA = CuSparseMatrixCSR(A)
-        dB = CuArray(B)
-        dC = CuArray(C)
-
-        mul!(C, A, B, alpha, beta)
-        mul!(dC, dA, dB, alpha, beta)
-        @test C ≈ collect(dC)
+                    C = opa(A) * opb(B)
+                    dC = opa(dA) * opb(dB)
+                    @test C ≈ collect(dC)
+                end
+            end
+        end
+
+        @testset "$SparseMatrixType * CuMatrix -- mul!(C, A, B) $elty" for elty in [Float32, Float64, ComplexF32, ComplexF64]
+            for opa in (identity, transpose, adjoint)
+                for opb in (identity, transpose, adjoint)
+                    CUSPARSE.version() < v"12.0" && SparseMatrixType == CuSparseMatrixCSC && elty <: Complex && opa == adjoint && continue
```

Comment:
is that because the functionality is broken on 12, or deprecated? adding a comment on that would be useful (so that we know what to do with these tests in the future)

Reply:
In CUDA toolkit < v"12.0", matrix-vector and matrix-matrix products  with the CuSparseMatrixCSC format was not implemented or fully supported (bugs).
We used the relation with the CuSparseMatrixCSR format to perform the operations.

The issue is that CSC format is the CSR format of the transposed matrix but when the sparse matrix has complex values we also need to conjugate the matrix to have the equivalence.
And because we don't have  a `trans` in CUSPARSE such as `N`, `T`, `C` to just perform the operation with the conjugate, the operation was deactivated.

I already added comments for that :wink: 
https://github.com/JuliaGPU/CUDA.jl/blob/tb/cuda_12/lib/cusparse/generic.jl#L144-L158
https://github.com/JuliaGPU/CUDA.jl/blob/tb/cuda_12/lib/cusparse/generic.jl#L200-L215

Reply:
A comment in the tests can't hurt :-) That said, I had missed the `continue`, so read this as tests only being enabled on CUDA 12+, which confused me.